### PR TITLE
block: add aio backend for raw disks.

### PR DESCRIPTION
**Pull request by thomasbarrett**
Some commonly used enterprise Linux distributions (RHEL 8 and RHEL 9)  do not support `io_uring`, despite having a sufficiently recent kernel. The current synchronous IO fallback has poor performance compared to `io_uring`. This PR adds a Linux aio backend as a fallback when `io_uring` is not supported. This improves block IO performance greatly on RHEL 8 and 9 when passing a raw block device to a virtual machine via virtio-blk. This PR always enables a fallback chain of `io_uring` -> `aio` -> `pread/pwrite`.

## Benchmarking
To test this change, I created a virtual machine with the following configuration:
```
cloud-hypervisor \
--kernel /var/lib/cloud-hypervisor/CLOUDHV.fd \
--cpus boot=20,topology=2:10:1:1,affinity=[0@[2],2@[3],4@[4],6@[5],8@[6],10@[7],12@[8],14@[9],16@[10],18@[11],1@[14],3@[15],5@[16],7@[17],9@[18],11@[19],13@[20],15@[21],17@[22],19@[23]] \
--memory size=0 \
--memory-zone size=32G,hugepages=on,hugepage_size=1G,prefault=on,host_numa_node=0,id=mem0 \
--numa guest_numa_id=0,cpus=[0-19],memory_zones=[mem0] \
--disk path=~/ubuntu.raw \
--disk path=/dev/nvme0n1,direct=on,queue_size=512 \
--console off \
--serial off \
--net "tap=" \
-v
```

Note that `nvme0n1` is a 1 TiB Samsung 980PRO NVMe drive. To benchmark this change, I used the following fio configuration. 
```
[global]
ioengine=libaio
direct=1
bs=4K
rw=randread
iodepth=512
size=100204886016
cpus_allowed=0

[job1]
filename=/dev/vdb
```
* When using the aio asynchronous backend, I benchmarked 473k IOPs and 1849MiB/s.
* When using the synchonrous backend, I benchmarked 27.4k IOPs and 107MiB/s.

### CODE REVIEW

**Comment by thomasbarrett** [comment]
Still unsure what the best UX for enabling aio is, but moving this to ready for review.

File: vmm/src/seccomp_filters.rs [hunk 1]
```diff
@@ -636,6 +636,11 @@ fn vmm_thread_rules(
         (libc::SYS_wait4, vec![]),
         (libc::SYS_write, vec![]),
         (libc::SYS_writev, vec![]),
+        (libc::SYS_io_submit, vec![]),
+        (libc::SYS_io_cancel, vec![]),
+        (libc::SYS_io_getevents, vec![]),
+        (libc::SYS_io_setup, vec![]),
+        (libc::SYS_io_destroy, vec![]),
```
Comment (likebreath) [inline vmm/src/seccomp_filters.rs:643]: Please follow alphabetical order here.

File: vmm/src/device_manager.rs [hunk 2]
```diff
@@ -2293,8 +2293,8 @@ impl DeviceManager {
                             Box::new(RawFileDisk::new(file)) as Box<dyn DiskFile>
                         }
                     } else {
-                        info!("Using synchronous RAW disk file");
-                        Box::new(RawFileDiskSync::new(file)) as Box<dyn DiskFile>
+                        info!("Using asynchronous RAW disk file (aio)");
+                        Box::new(RawFileDiskAio::new(file)) as Box<dyn DiskFile>
```
Comment (likebreath) [inline vmm/src/device_manager.rs:2297]: Can you add a check for the support of aio on the host at runtime (like what's done from `io_uring_is_supported()` and keep the sync I/O as default?

**Review by likebreath — COMMENTED** [review]
I'd keep a fall back chain of `io_uring` -> `aio` -> `sync io` by default, while providing (non-public) options to manually disable `io_uring` and/or `aio` (for testing purposes or niche use-cases only). See this commit for reference: https://github.com/cloud-hypervisor/cloud-hypervisor/commit/c62227803047d0bf81279cb4aadcf131a194426d.

Also, can you please add a integration test for the aio backend? You may can reuse `_test_virtio_block()`.

File: vmm/src/device_manager.rs (hunk 2 above)
Comment (thomasbarrett) [reply vmm/src/device_manager.rs:2312]: Done. Unfortunately, aio doesn't have a comprehensive probing system like io_uring, but the check I added should be sufficient to detect cases where aio has been explicitly compiled out of the kernel.

File: vmm/src/seccomp_filters.rs (hunk 1 above)
Comment (thomasbarrett) [reply vmm/src/seccomp_filters.rs]: Updated to follow alphabetical order.

**Comment by thomasbarrett** [comment]
Yeah I really like the `io_uring` -> `aio` -> `sync` fallback chain idea @likebreath. Updated. On a side note, the `aarch64` CI runner seems to be having some issues.

**Review by likebreath — APPROVED** [review]
Thank you for the prompt response. Good work.

File: block/src/raw_async_aio.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
+use crate::async_io::{
```
Comment (rbradford) [inline block/src/raw_async_aio.rs:1]: This new file hasn't got a copyright header.

File: vmm/src/config.rs [hunk 4]
```diff
@@ -928,6 +929,11 @@ impl DiskConfig {
             .map_err(Error::ParseDisk)?
             .unwrap_or(Toggle(false))
             .0;
+        let disable_aio = parser
```
Comment (rbradford) [inline vmm/src/config.rs:932]: We don't really need this - we only added `_disable_io_uring` as we supported kernels that didn't have it and we needed some way to test. Our minimum kernel now does support io_uring so we can remove this at some point.

**Review by rbradford — COMMENTED** [review]
Please address the missing copyright header on the added file as soon as possible.

File: block/src/raw_async_aio.rs (hunk 3 above)
Comment (thomasbarrett) [reply block/src/raw_async_aio.rs:1]: Thanks for looking back at this @rbradford. Put out a PR to add.

https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5888

**Pull request merged**